examples/qoi.cpp

```diff
@@ -0,0 +1,251 @@
+
+    mfem::BlockVector rhs_fine(upscale.GetBlockVector(0));
+    rhs_fine.GetBlock(0) = 0.0;
+    rhs_fine.GetBlock(1) = 0.0;
+    // it may make sense to have more general, not hard-coded, right-hand side
+    int index = 0.5 * scale * scale + 0.1 * scale;
+    rhs_fine.GetBlock(1)(index) = 1.0;
+    index = 0.6 * scale * scale + 0.75 * scale;
+    rhs_fine.GetBlock(1)(index) = -1.0;
+
+    std::unique_ptr<MultilevelSampler> sampler;
+    const int seed = argseed + myid;
+    sampler = make_unique<PDESampler>(
+                  dimension, cell_volume, kappa, seed,
+                  graph, upscale_param);
+
+    mfem::Vector functional(rhs_fine.GetBlock(1));
+    functional = 0.0;
+    double dscale = 1.0 / scale;
+    for (int i = 0; i < scale; ++i)
+    {
+        functional(i) = dscale;
+    }
+    mfem::Vector dummy;
+    PressureFunctionalQoI qoi(upscale, functional);
+
+    MLMCManager mlmc(*sampler, qoi, upscale, rhs_fine, dump_number);
+
+    const bool verbose = (myid == 0);
+
+    if (upscale_param.max_levels == 1)
+    {
+        for (int sample = 0; sample < fine_samples; ++sample)
+        {
+            if (myid == 0)
+                std::cout << "---\nFine sample " << sample << "\n---" << std::endl;
+
+            mlmc.FineSample(verbose);
+        }
+    }
+    else if (upscale_param.max_levels == 2)
+    {
+        for (int sample = 0; sample < coarse_samples; ++sample)
+        {
+            if (myid == 0)
+                std::cout << "---\nCoarsest " << upscale_param.max_levels - 1 << " sample "
+                          << sample << "\n---" << std::endl;
+            mlmc.CoarseSample(verbose);
+        }
+        for (int sample = 0; sample < shared_samples; ++sample)
+        {
+            if (myid == 0)
+                std::cout << "---\nShared sample " << sample << "\n---" << std::endl;
+
+            mlmc.CorrectionSample(0, verbose);
+        }
+    }
+    else
+    {
+        // at least two samples on each level to ensure a meaningful variance (and cost)
+        // three is better; sometimes you get stuck if there are very few samples on a level
+        for (int i = 0; i < coarse_samples; ++i)
+        {
+            std::cout << "---\nCoarsest " << upscale_param.max_levels - 1 << " sample "
+                      << i << "\n---" << std::endl;
+            mlmc.CoarseSample(verbose);
+        }
+        for (int level = 0; level < upscale_param.max_levels - 1; ++level)
+        {
+            for (int i = 0; i < shared_samples; ++i)
+            {
+                std::cout << "---\nInitial level " << level << " correction " << i
+                          << "\n---" << std::endl;
+                mlmc.CorrectionSample(level, verbose);
+            }
+        }
+    }
+
+    for (int sample = 0; sample < choose_samples; ++sample)
+    {
+        if (myid == 0)
+            std::cout << "---\nChoose sample " << sample << "\n---" << std::endl;
+
+        mlmc.BestSample(verbose);
+    }
```

Comment:
It probably makes sense to wrap L189-243 into a function 
(say, `MLMCManager::Simulate(int num_levels, bool verbose)`, where `num_levels` is the number of levels of the MLMC simulation, which can be different from the number of levels of `Upscale` object that MLMCManager is constructed from)